src/lib/auth.ts

```diff
@@ -0,0 +1,110 @@
+import * as vscode from 'vscode';
+import * as fs from 'fs';
+import { Portal } from './types';
+import { COMMANDS } from './constants';
+
+const { findConfig, loadConfig, validateConfig } = require('@hubspot/cli-lib');
+const { setConfig, setConfigPath } = require('@hubspot/cli-lib/lib/config');
+
+let hubspotConfigWatcher: fs.FSWatcher | null;
+let configDependentFeaturesLoaded: boolean = false;
+let loadConfigDependentFeatures: Function;
+let updateConfigDependentFeatures: Function;
+
+const onLoadPath = (configPath: string) => {
+  vscode.commands.executeCommand(
+    'setContext',
+    'hubspot.configPath',
+    configPath
+  );
+  if (!configPath) {
+    vscode.commands.executeCommand(COMMANDS.CONFIG_SET_DEFAULT_ACCOUNT, null);
+    setConfig(null);
+    setConfigPath(null);
+  }
+};
+
+const onLoadHubspotConfig = () => {
+  updateConfigDependentFeatures();
+};
+
+export const loadHubspotConfigFile = (rootPath: string) => {
+  if (!rootPath) {
+    return;
+  }
+
+  console.log(`Root path: ${rootPath}`);
+
+  const path = findConfig(rootPath);
+  onLoadPath(path);
+
+  console.log(`Path: ${path}`);
+
+  if (!path) {
+    return;
+  }
+
+  const config = loadConfig(path);
+
+  console.log(
+    'loadedHubspotConfig',
+    config.defaultPortal,
+    JSON.stringify(config.portals.map((p: Portal) => p.name))
+  );
+
+  if (!validateConfig()) {
+    throw new Error(`Invalid config could not be loaded: ${path}`);
+  } else {
+    onLoadHubspotConfig();
+    return path;
+  }
+};
+
+export const initializeHubspotConfigDependents = (
+  rootPath: string,
+  configPath: string
+) => {
+  if (!configDependentFeaturesLoaded) {
+    configDependentFeaturesLoaded = true;
+    loadConfigDependentFeatures();
+  }
+
+  // This triggers an in-memory update of the config when the file changes
+  if (!hubspotConfigWatcher) {
+    console.log('Started watching hubspot.config.yml');
+    hubspotConfigWatcher = fs.watch(configPath, async (eventType) => {
+      if (eventType === 'change') {
+        console.log('hubspot.config.yml changed');
+        loadHubspotConfigFile(rootPath);
+      } else if (eventType === 'rename') {
+        console.log('hubspot.config.yml renamed/deleted');
+        loadHubspotConfigFile(rootPath);
+        hubspotConfigWatcher && hubspotConfigWatcher.close();
+        hubspotConfigWatcher = null;
+        console.log('Stopped watching hubspot.config.yml');
+      }
+    });
+  }
+};
+
+export const registerConfigDependentFeatures = async ({
+  rootPath,
+  onConfigFound,
+  onConfigUpdated,
+}: {
+  rootPath: string;
+  onConfigFound: Function;
+  onConfigUpdated: Function;
+}) => {
+  loadConfigDependentFeatures = onConfigFound;
+  updateConfigDependentFeatures = onConfigUpdated;
```

Comment:
I might not be following this correctly but are we hoisting these functions to later be called by `initializeHubspotConfigDependents()`? If so, could they get defined in the same scope and called directly?

Reply:
This is clearly more confusing than it needs to be. I refactored the callbacks into `vscode.command`s. I think it should be an improvement.